From 0e18a4deb6f2b8e7a408676cb74b6117e9c33351 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:51:37 -0400
Subject: [fs] nfsd: Error out when getting more than one fsloc/secinfo/uuid

Message-id: <1403560314-32525-156-git-send-email-bfields@redhat.com>
Patchwork-id: 84581
O-Subject: [Patch RHEL7 155/172] NFSD: Error out when getting more than one fsloc/secinfo/uuid
Bugzilla: 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Kinglong Mee <kinglongmee@gmail.com>

Signed-off-by: Kinglong Mee <kinglongmee@gmail.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: be69da8052af38a9d72ed32fb765fd4446e4091c
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/export.c b/fs/nfsd/export.c
index 7884051..13b85f9 100644
--- a/fs/nfsd/export.c
+++ b/fs/nfsd/export.c
@@ -389,6 +389,10 @@ fsloc_parse(char **mesg, char *buf, struct nfsd4_fs_locations *fsloc)
  int len;
  int migrated, i, err;
 
+ /* more than one fsloc */
+ if (fsloc->locations)
+  return -EINVAL;
+
  /* listsize */
  err = get_uint(mesg, &fsloc->locations_count);
  if (err)
@@ -442,6 +446,10 @@ static int secinfo_parse(char **mesg, char *buf, struct svc_export *exp)
  u32 listsize;
  int err;
 
+ /* more than one secinfo */
+ if (exp->ex_nflavors)
+  return -EINVAL;
+
  err = get_uint(mesg, &listsize);
  if (err)
   return err;
@@ -481,6 +489,10 @@ uuid_parse(char **mesg, char *buf, unsigned char **puuid)
 {
  int len;
 
+ /* more than one uuid */
+ if (*puuid)
+  return -EINVAL;
+
  /* expect a 16 byte uuid encoded as \xXXXX... */
  len = qword_get(mesg, buf, PAGE_SIZE);
  if (len != EX_UUID_LEN)
-- 
1.7.1